src/track/beats.cpp

```diff
@@ -25,40 +30,22 @@ inline FrameDiff_t getBeatLengthFrames(Bpm bpm, double sampleRate) {
 Beats::Beats(const Track* track, const QVector<FramePos>& beats)
         : Beats(track) {
     if (beats.size() > 0) {
-        FramePos previous_beatpos = kInvalidFramePos;
-        track::io::Beat protoBeat;
-
-        for (auto beat : beats) {
-            if (beat <= previous_beatpos || beat < FramePos(0)) {
-                qDebug() << "kBeatMap::createFromVector: beats not in increasing order or negative";
-                qDebug() << "discarding beat " << beat;
-            } else {
-                protoBeat.set_frame_position(beat.getValue());
-                m_beatsInternal.m_beats.append(protoBeat);
-                previous_beatpos = beat;
-            }
-        }
-        updateBpm();
+        // This causes BeatsInternal constructor to be called twice.
+        // But it can't be included in ctor initializer list since
+        // we already have a delegating constructor.
+        m_beatsInternal = BeatsInternal(beats);
     }
+    slotTrackBeatsUpdated();
 }
 
 Beats::Beats(const Track* track, const QByteArray& byteArray)
         : Beats(track) {
-    track::io::Beats beatsProto;
-    if (!beatsProto.ParseFromArray(byteArray.constData(), byteArray.size())) {
-        qDebug() << "ERROR: Could not parse kBeatMap from QByteArray of size"
-                 << byteArray.size();
-    }
-    for (int i = 0; i < beatsProto.beat_size(); ++i) {
-        const track::io::Beat& beat = beatsProto.beat(i);
-        m_beatsInternal.m_beats.append(beat);
-    }
-    updateBpm();
+    m_beatsInternal = BeatsInternal(byteArray);
+    slotTrackBeatsUpdated();
 }
 
 Beats::Beats(const Track* track)
-        : m_mutex(QMutex::Recursive),
-          m_track(track) {
+        : m_mutex(QMutex::Recursive), m_track(track) {
```

Comment:
Did clang-format do this? We put each item of the initializer list on separate lines.

Reply:
Probably. I did not touch this constructor.